synapse/storage/databases/main/roommember.py

```diff
@@ -1572,6 +1572,37 @@ def get_sliding_sync_room_for_user_batch_txn(
             get_sliding_sync_room_for_user_batch_txn,
         )
 
+    async def get_rooms_for_user_by_date(self, user_id: str, from_ts: int) -> frozenset:
+        """
+        Fetch a list of rooms that the user has joined since the given timestamp.
+
+        Args:
+            user_id: user ID of the user to search for
+            from_ts: a timestamp in ms from the unix epoch at which to begin the search at
+        """
+
+        def _get_rooms_for_user_by_join_date_txn(
+            txn: LoggingTransaction, user_id: str, timestamp: int
+        ) -> frozenset:
+            # Note that matching on c.type = 'm.room.member' allows us to benefit from
+            # this index: "current_state_events_member_index" btree (state_key) WHERE type = 'm.room.member'::text
+            # on the current_state_events_table
+            sql = """
+                    SELECT c.room_id
+                    FROM current_state_events c
+                    JOIN events e ON c.event_id = e.event_id
+                    WHERE c.state_key = ? AND c.membership = 'join' AND e.received_ts > ? AND c.type = 'm.room.member'
```

Comment:
We at-least need to match on `e.type` (not `c.type`) to use the new index

```suggestion
                    WHERE c.state_key = ? AND c.membership = 'join' AND e.received_ts > ? AND e.type = 'm.room.member'
```

Reply:
I am slightly confused by this - the new index is on `events.received_ts` - just wondering how matching on another column takes advantage of that index?

Reply:
The index is on `events.received_ts` but only when `events.type = 'm.room.member'` (the `where_clause` makes it a partial index).

If we match on `current_state_events.type = 'm.room.member'`, the SQL engine might not be smart enough to know we can use the **`events.type = 'm.room.member'`** partial index. `current_state_events.type` vs `events.type` are their own tables with their own columns.

Reply:
Ah that makes sense, thanks for explaining.